src/name_utils.rs

```diff
@@ -0,0 +1,399 @@
+    let selectable_chars = selectable_chars(&sub_names, max_chars);
+    // The `carryover` are left over characters that will not be used by some sub names
+    // and can be added to other longer sub names exceeding the amount of `selectable_chars`.
+    let carryover = max_chars - used_characters(&sub_names, selectable_chars);
+
+    Ok(build_name(
+        sub_names.as_slice(),
+        selectable_chars,
+        carryover,
+    ))
+}
+
+/// This splits the number of available characters into equal blocks for each sub name and
+/// returns its size.
+///
+/// # Arguments
+///
+/// * `sub_names` - A vector of available sub names.
+/// * `max_chars` - The maximum available characters:
+///                 `RESOURCE_NAME_MAX_LEN` - `KUBERNETES_HASH_MIN_LENGTH` - sub_names.len()
+///
+fn selectable_chars(sub_names: &[String], max_chars: usize) -> usize {
+    (1f32 / sub_names.len() as f32 * max_chars as f32).floor() as usize
+}
+
+/// This calculates the sum of used characters from each sub name. Sub names that exceed the
+/// amount of `selectable_chars` will count as the number of `selectable_chars`.
+///
+/// # Arguments
+///
+/// * `sub_names` - A vector of available sub names.
+/// * `selectable_chars` - Even distributed amount of characters that may be used for each sub name.
+///
+fn used_characters(sub_names: &[String], selectable_chars: usize) -> usize {
+    let mut used = 0;
+
+    for name in sub_names {
+        if selectable_chars > name.len() {
+            used += name.len();
+        } else {
+            used += selectable_chars;
+        }
+    }
+
+    used
+}
+
+/// This method concatenates each sub name and dynamically adapts the length of each sub name if
+/// `unused_chars` are available. The order of 'distributing' the unused_chars is determined by
+/// the sub_names vector. Items that come first may receive all `unused_chars` while items in
+/// the end have to be cut down to the length of `selectable_chars`.
+///
+/// # Arguments
+///
+/// * `sub_names` - A vector of available sub names.
+/// * `selectable_chars` - Even distributed amount of characters that may be used for each sub name.
+/// * `unused_chars` - Number of characters that are not used by some sub names and may be added to
+///                    other longer sub names.
+///
+fn build_name(sub_names: &[String], selectable_chars: usize, unused_chars: usize) -> String {
+    let mut full_name = String::new();
+    let mut carryover = unused_chars;
+
+    for name in sub_names {
+        let selected: String = name
+            .chars()
+            .into_iter()
+            .take(selectable_chars + carryover)
+            .collect();
+
+        // if the sub name was extended via carryover, we need to adapt how many
+        // carryover characters we have left.
+        if selected.len() > selectable_chars {
+            carryover -= selected.len() - selectable_chars;
+        }
+
+        full_name.push_str(&selected);
+        full_name.push('-');
```

Comment:
This always adds a '-' at the end. Not  good. Use a `Vec<String>` for `full_name` and do `full_name.join("-")` at the end.

Reply:
Yes thats what we want. It is intended to work with generatedName from kubernetes, which will append a hash at the end. Wont look good without that.